0010-B-object-storage-building-block.md

```diff
@@ -0,0 +1,275 @@
+
+  // SetObject stores an object with name key.
+  rpc SetObject(stream SetObjectRequest) returns (SetObjectResponse) {}
+
+  // DeleteObject an object with name key.
+  rpc DeleteObject(DeleteObjectRequest) returns (DeleteObjectResponse) {}
+}
+
+// GetObjectRequest is the message to get an object from specific object store.
+message GetObjectRequest {
+  // The name of object store.
+  string store_name = 1;
+
+  // The key of the desired object.
+  string key = 2;
+}
+
+// GetObjectResponse contains the retrieved object (this is used in a streamed response).
+message GetObjectResponse {
+  // The tag for concurrency control.
+  string tag = 1;
+
+  // The metadata stored with the object.
+  map<string, string> metadata = 2;
+
+  // Chunk of data.
+  StreamPayload payload = 10;
+}
+
+// SetObjectRequest is the message to store a object in a specific object store (this is used in a streamed request).
+message SetObjectRequest {
+  // The name of object store.
+  string store_name = 1;
+
+  // The key of the desired object.
+  string key = 2;
+
+  // The tag for concurrency control.
+  string tag = 3;
+
+  // The metadata which will be stored with the object.
+  map<string, string> metadata = 4;
+
+  // Chunk of data.
+  StreamPayload payload = 10;
+}
+
+// SetObjectResponse contains the result of storing the object.
+message SetObjectResponse {
+  // The updated tag for concurrency control.
+  string tag = 1;
+
+  // The number of bytes written.
+  int bytes = 2;
+}
+
+// DeleteObjectRequest is the message to delete an object in specific object store.
+message DeleteObjectRequest {
+  // The name of object store.
+  string store_name = 1;
+
+  // The key of the desired object.
+  string key = 2;
+}
+
+// DeleteObjectResponse contains the result of deleting the object.
+message DeleteObjectResponse {
+    // Currently empty but allowing for future expansion.
+}
+```
+
+#### Handling streams
+
+`StreamPayload` is first introduced with dapr/dapr#4903 and corresponds to:
+
+```proto3
+// Chunk of data sent in a streaming request or response.
+message StreamPayload {
+  // Data sent in the chunk.
+  google.protobuf.Any data = 1;
+
+  // Set to true if this is the last chunk.
+  bool complete = 2;
+}
+```
+
+Data is sent from the application to Dapr (in `SetObject` RPCs) or from Dapr to the application (in `GetObject` RPCs) in a stream. Each message in the stream contains a `GetObjectResponse` or `SetObjectRequest` where:
+
+- The first message in the stream MUST contain all other required keys.
+- The first message in the stream MAY contain a `payload`, but that is not required.
+- Subsequent messages (any message except the first in the stream) MUST contain a `payload` and MUST NOT contain any other property.
+- The last message in the stream MUST contain a `payload` with `complete` set to `true`. That message is assumed to be the last one from the sender and no more messages are to be sent in the stream after that.
```

Comment:
See above, but I'd also recommend being able to send any other metadata at the end. For example, content-length, content-hash, or whatever else is important to the application that can only be done once the entire file is consumed.

Reply:
I will let the server populate content-hash and/or content-length.

For checksumming during upload, if the service and SDK support that, it can be done transparently by the component. Because components will most likely need to performed chunked upload, the checksum should be included in each chunk anyways.